File: Linked Lists/2130. Maximum Twin Sum of a Linked List/2130. Maximum Twin Sum of a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        fast = head.next.next
        slow = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        prev = None
        slow = slow.next
        while (slow):
            temp = slow.next
            slow.next = prev
            prev, slow = slow, temp
        
        max_sum = 0
        left = head
        right = prev
        while (right):
            max_sum = max(max_sum, left.val + right.val)
            left = left.next
            right = right.next
        
        return max_sum

# Time: O(n)
# Space: O(1)

File: Linked Lists/2130. Maximum Twin Sum of a Linked List/2130. Maximum Twin Sum of a Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number}
 */
var pairSum = function (head) {
  let fast = head.next.next;
  let slow = head;

  while (fast && fast.next) {
    fast = fast.next.next;
    slow = slow.next;
  }

  let prev = null;
  slow = slow.next;
  while (slow) {
    const temp = slow.next;
    slow.next = prev;
    prev = slow;
    slow = temp;
  }

  let max_sum = 0;
  let left = head;
  let right = prev;
  while (right) {
    max_sum = Math.max(max_sum, left.val + right.val);
    left = left.next;
    right = right.next;
  }

  return max_sum;
};

// Time: O(n)
// Space: O(1)
